zetaclient/evm_client.go

```diff
@@ -756,36 +768,56 @@ func (ob *EVMChainClient) checkConfirmedTx(txHash string, nonce uint64) (*ethtyp
 		return nil, nil, false
 	}
 
-	// cross-check receipt against the block
-	block, err := ob.GetBlockByNumberCached(receipt.BlockNumber.Uint64())
-	if err != nil {
-		log.Error().Err(err).Msgf("confirmTxByHash: GetBlockByNumberCached error, txHash %s nonce %d block %d",
-			txHash, nonce, receipt.BlockNumber)
-		return nil, nil, false
-	}
-	// #nosec G701 non negative value
-	if receipt.TransactionIndex >= uint(len(block.Transactions())) {
-		log.Error().Msgf("confirmTxByHash: transaction index %d out of range [0, %d), txHash %s nonce %d block %d",
-			receipt.TransactionIndex, len(block.Transactions()), txHash, nonce, receipt.BlockNumber)
-		return nil, nil, false
-	}
-	txAtIndex := block.Transactions()[receipt.TransactionIndex]
-	if txAtIndex.Hash() != transaction.Hash() {
-		log.Error().Msgf("confirmTxByHash: transaction at index %d has different hash %s, txHash %s nonce %d block %d",
-			receipt.TransactionIndex, txAtIndex.Hash().Hex(), txHash, nonce, receipt.BlockNumber)
-		return nil, nil, false
-	}
-
 	// check confirmations
 	if !ob.HasEnoughConfirmations(receipt, ob.GetLastBlockHeight()) {
 		log.Debug().Msgf("confirmTxByHash: txHash %s nonce %d included but not confirmed: receipt block %d, current block %d",
 			txHash, nonce, receipt.BlockNumber, ob.GetLastBlockHeight())
 		return nil, nil, false
 	}
 
+	// cross-check tx inclusion against the block
+	// Note: a guard for false BlockNumber in receipt. The blob-carrying tx won't come here
+	err = ob.checkTxInclusion(transaction, receipt.BlockNumber.Uint64(), receipt.TransactionIndex)
+	if err != nil {
+		log.Error().Err(err).Msgf("confirmTxByHash: checkTxInclusion error for txHash %s nonce %d", txHash, nonce)
+		return nil, nil, false
+	}
+
 	return receipt, transaction, true
 }
 
+// checkTxInclusion returns nil only if tx is included in the block at blockNumber and txIndex
+func (ob *EVMChainClient) checkTxInclusion(tx *ethtypes.Transaction, blockNumber uint64, txIndex uint) error {
```

Comment:
Can we define a new type instead of using a bool for fallback?
If so, we can use the switch instead of the `if, else` 
The code would look cleaner IMO , but don't feel too strongly about it

Reply:
do you mean switch on types or something?

Reply:
I meant something like this 
`type EthRPC int64
	var (
		regular  EthRPC = 1
		fallback EthRPC = 2
	)`